### PR TITLE
Rebuild as a content storage service

### PR DESCRIPTION
**Pull request by jayzes**
# Rendering Service
## Goals
- Persistent cache of rendered content in order to mitigate memory only cache failure and resulting load/latency spike.
- Ability to add new pipeline properties and pre-cache without adding huge load/latency spike.
- Less then 50ms of added latency to majority of post rendering requests in API (equvilent to existing memcached calls).
- Bonus: Ability to render non post-content blocks (such as eventual PMs).
## Usage:

In an index action (or anywhere we render lists of posts):

``` ruby
posts.tap {|posts| PostRenderer.prefetch(posts) }
```

In single post scenarios no updates required.
### Bulk Fetch from cache:

Used to get a list of block/content bodies based on their checksum. Returns an array of exactly the same size as was passed in with null entries for cache misses.

POST /v1/bulk_fetch

Body:

```
[{checksum: md5, pipeline: "pipelinename"}, ... ]
```

Returns:

```
[{rendered_content: ""}]
```
### Render

Returns the rendered content either for a stored version based on md5 or a new version based on the content passed in.

POST /v1/render

body:

```
{checksum: md5, pipeline: "pipelinename", content: ""}
```

Returns:

```
{rendered_content: ""}
```
## Storage

**original_contents**

Stores original version of the content. When new "pipeline" is added this is the source. Given multiple context's this will have multiple records.
- id
- content_checksum
- original_content

**rendered_contents**

Stores a rendered version of the content for a given pipeline. Each version of the pipeline would result in full new set of records in this table.
- id
- original_content_id
- pipeline - string representation of the pipeline name
- rendered_content

TODOs:
- [x] Bring back `rack-timeout` and `rails-12factor`
- [x] Split out `v1_content`, `v1_bio`, `v1_summary_120` and `v1_summary_160` pipelines
- [x] Add basic auth
- [x] Unique index on checksums
- [x] Short-circuit render if we have a matching checksum
- [x] Dockerize and add to mothership compose.yml
- [x] Performance test and benchmark
- [x] Deal with `sanitize` calls
- [ ] Benchmark performance for cache misses and figure out if they need to be backgrounded somehow

### CODE REVIEW

**Comment by alanpeabody** [comment]
:shipit:

**Comment by jayzes** [comment]
:trollface:

**Comment by alanpeabody** [comment]
Still 👍

File: spec/requests/bulk_fetch_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+           params: [
+             { checksum: checksum,
+               pipeline: pipeline }
+           ].to_json,
+           headers: { 'Content-Type' => 'application/json' }
+    end
+
+    it 'returns a rendered form of the content' do
+      expect(response_json).to eq(
+        [ { 'rendered_content' => '<h3>Hello, world!</h3>' } ]
```
Comment (alanpeabody) [inline spec/requests/bulk_fetch_spec.rb]: I think we are actually going to need the checksum and pipeline returned in the response as well.

File: spec/requests/render_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+         params: {
+           checksum: checksum,
+           pipeline: pipeline,
+           content: content
+         }.to_json,
+         headers: { 'Content-Type' => 'application/json' }
+  end
+
+  it 'returns a rendered form of the content' do
+    expect(response.body).to eq('<h3>Hello, world!</h3>')
```
Comment (alanpeabody) [inline spec/requests/render_spec.rb]: Could we return json here just to have a standard response? Pretty much the same as the other endpoint but just a single object?

File: config/initializers/render_pipeline.rb [hunk 3]
```diff
@@ -20,7 +20,27 @@
   config.add_emoji 'pride'
   config.add_emoji 'hot_shit'
 
-  config.render_context :default do |c|
+  config.render_context :default, :v1_bio, :v1_content do |c|
```
Comment (alanpeabody) [inline config/initializers/render_pipeline.rb]: Does Bio need to truncate?

Comment (jayzes) [inline config/initializers/render_pipeline.rb]: It truncates at 666 currently. I don't think post content does, however, so we might want to separate those two.

File: app/interactors/render_content.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+class RenderContent
+  include Interactor
+
+  def call
+    source = SourceContentItem.
+             where(checksum: context.checksum).
+             first_or_initialize(content: context.content)
+    render_content_for(source) if source.new_record?
+    context.fail! unless source.save
```
Comment (alanpeabody) [inline app/interactors/render_content.rb]: We only need to save this if it is a new record right?

**Comment by alanpeabody** [comment]
LGTM!

**Comment by steam** [comment]
@jayzes You want to take a final look at this?

**Comment by jayzes** [comment]
@steam looks good to me! good catch on the empty JSON case, I didn't think about that one.

**Comment by steam** [comment]
The empty json wasn't obvious at all. We shipped the changes to staging and promptly saw out of memory errors. We increased the size of the dyno and continued to see memory errors. The `WHERE` clause was omitted in the case of empty params. Once we accounted for empty json the response times went WAY down 8,000ms+ down to under 20ms or so.

**Comment by steam** [comment]
Last note: @alanpeabody is going to make sure the mothership does not attempt to talk to the render service in the case where no regions are present. That way the empty `json` params are not passed in the first place.

**Comment by alanpeabody** [comment]
Addressed in: https://github.com/ello/ello/pull/1780/commits/a6868835310ecc6d0767be14e814e00305de855d

**Pull request closed without merging**